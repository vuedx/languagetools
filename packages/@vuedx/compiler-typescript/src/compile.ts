import {
  createCallExpression,
  CREATE_BLOCK,
  CREATE_VNODE,
  FRAGMENT,
  generate,
  OPEN_BLOCK,
  trackSlotScopes,
  trackVForSlotScopes,
  transform,
  TransformContext,
  transformExpression,
  RootNode,
  CodegenResult,
} from '@vue/compiler-core';
import { transformElement } from './nodes/transformElement';
import { transformFor } from './nodes/transformFor';
import { transformIf } from './nodes/transformIf';
import { transformSlotOutlet } from './nodes/transformSlotOutlet';
import { pascalCase } from './pascalCase';
import { H } from './runtimeHelpers';
import { Options } from './options';
import { once, cloak, show, text, on, bind, model } from './directives';

export function compile(ast: RootNode, options: Options): CodegenResult {
  let alreadyExecuted = false;
  let context!: TransformContext;
  transform(ast, {
    ...options,
    prefixIdentifiers: true,
    hoistStatic: false,
    transformHoist: null,
    ssr: false,
    cacheHandlers: false,
    scopeId: null,
    nodeTransforms: [
      transformIf,
      transformFor,
      trackVForSlotScopes,
      transformExpression,
      transformSlotOutlet,
      transformElement(Object.keys(options.components).map(pascalCase)),
      trackSlotScopes,
      ...(options.nodeTransforms || []),
      (_, _context) => {
        if (alreadyExecuted) return;
        context = _context;
        alreadyExecuted = true;
        context.helper(FRAGMENT);
        context.imports.add({ exp: '_Ctx', path: options.filename });
        Object.entries(options.components).forEach(([name, path]) => {
          context.imports.add({
            exp: `_component_${pascalCase(name)}`,
            path: path,
          });
        });
      },
    ],
    directiveTransforms: {
      once,
      cloak,
      show,
      text,
      on,
      bind,
      model,
      ...options.directiveTransforms,
    },
  });
  ast.codegenNode = createCallExpression(context.helperString(context.helper(H)), [
    context.helperString(context.helper(FRAGMENT)),
    ast.children,
  ]);
  [OPEN_BLOCK, CREATE_BLOCK, CREATE_VNODE].forEach((helper) => {
    const index = ast.helpers.indexOf(helper);
    if (index >= 0) ast.helpers.splice(index, 1);
  });
  const result = generate(ast, {
    ...options,
    mode: 'module',
    optimizeBindings: false,
    prefixIdentifiers: true,
    scopeId: null,
    ssr: false,
    sourceMap: true,
  });
  result.code = result.code.replace(
    `export function render(_ctx, _cache)`,
    options.useJavaScript
      ? `export /** @param {InstanceType<typeof _Ctx>} _ctx */ function render(_ctx)`
      : `export function render(_ctx: InstanceType<typeof _Ctx>)`
  );
  return result;
}
