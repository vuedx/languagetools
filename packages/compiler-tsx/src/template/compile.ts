import {
  CompilerError,
  CompilerOptions,
  RootNode,
  transform,
} from '@vue/compiler-core'
import { withScope } from './scope/Scope'
import type { NodeTransformContext } from './types/NodeTransformContext'
import { ScopeManager } from './scope/ScopeManager'
import { createComponentChildrenTransform } from './transforms/transformComponent'
import { createTransformFor } from './transforms/transformFor'
import { createTransformIf } from './transforms/transformIf'
import { createResolveComponentTransform } from './transforms/transformResolveComponent'
import { createSlotHoistScopeTransform } from './transforms/transformSlot'
import type { TransformedCode } from '../types/TransformedCode'
import type { TransformOptionsResolved } from '../types/TransformOptions'

import { generate } from './generate'
import { parse } from './parse'

export interface Options {
  fileName: string
}

export interface Output extends TransformedCode {
  ast: RootNode
  errors: CompilerError[]
}

export function compile(
  template: string,
  options: Options & CompilerOptions & TransformOptionsResolved,
): Output {
  const errors: CompilerError[] = []
  const result = compileFromAST(
    parse(template, {
      ...options,
      onError: (error) => {
        errors.push(error)
        options.onError?.(error)
      },
    }),
    options,
  )

  result.errors.unshift(...errors)

  return result
}

export function compileFromAST(
  source: RootNode,
  options: Options & TransformOptionsResolved,
): Output {
  const root = withScope(clone(source))
  const ast = clone(source)
  ast.scope = root.scope
  const context: NodeTransformContext = {
    ...options,
    scope: new ScopeManager(),
    used: { components: new Set(), directives: new Set() },
  }
  const errors: CompilerError[] = []

  transform(root, {
    ...options,
    prefixIdentifiers: false,
    hoistStatic: false,
    transformHoist: null,
    ssr: false,
    cacheHandlers: false,
    scopeId: null,
    nodeTransforms: [
      createResolveComponentTransform(context),
      createSlotHoistScopeTransform(context),
      createTransformFor(context),
      createTransformIf(context),
      createComponentChildrenTransform(context),
    ],
    onError(error) {
      errors.push(error)
    },
  })

  const result = generate(root, context)

  return { ...result, ast, errors }
}

function clone<T>(obj: T): T {
  // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error
  // @ts-ignore
  if (typeof structuredClone !== 'undefined') return structuredClone(obj)
  return JSON.parse(JSON.stringify(obj))
}
