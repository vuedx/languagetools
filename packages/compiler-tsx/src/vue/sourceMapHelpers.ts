import type { DecodedSourceMap } from 'magic-string'
import { decode } from 'sourcemap-codec'
import type { SourceMapLike } from '../types/SourceMapLike'

export function getMappings(
  sourceMap: SourceMapLike,
): DecodedSourceMap['mappings'] {
  return typeof sourceMap.mappings === 'string'
    ? decode(sourceMap.mappings)
    : sourceMap.mappings
}

export function rebaseSourceMap(
  sourceMap: SourceMapLike,
  startPosition?: { line: number; column: number },
): DecodedSourceMap {
  const mappings = getMappings(sourceMap)

  if (startPosition == null) {
    return {
      file: '',
      sources: [],
      sourcesContent: [],
      ...sourceMap,
      names: sourceMap.names ?? [],
      mappings,
    }
  }

  const line = startPosition.line - 1
  const column = startPosition.column - 1

  return {
    file: '',
    sources: [],
    sourcesContent: [],
    ...sourceMap,
    names: sourceMap.names ?? [],
    mappings: mappings.map((mapping) =>
      mapping.map((segment) => {
        if (segment.length === 1) return segment

        let originalLine = segment[2]
        let originalColumn = segment[3]

        if (originalLine === 0) {
          originalColumn += column
        }

        originalLine += line

        if (segment.length === 4) {
          return [segment[0], segment[1], originalLine, originalColumn]
        }

        return [
          segment[0],
          segment[1],
          originalLine,
          originalColumn,
          segment[4],
        ]
      }),
    ),
  }
}
