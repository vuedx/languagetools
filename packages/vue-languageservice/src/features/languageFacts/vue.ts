import type { HTMLDataV1 } from 'vscode-html-languageservice'

export const data: HTMLDataV1 = {
  version: 1.1,
  tags: [
    {
      name: 'template',
      attributes: [
        {
          name: 'lang',
          values: [{ name: 'html' }, { name: 'pug' }],
        },
      ],
    },
    {
      name: 'script',
      attributes: [
        {
          name: 'lang',
          values: [
            { name: 'js' },
            { name: 'ts' },
            { name: 'jsx' },
            { name: 'tsx' },
          ],
        },
        { name: 'setup', valueSet: 'v' },
      ],
    },
    {
      name: 'style',
      attributes: [
        {
          name: 'lang',
          values: [
            { name: 'css' },
            { name: 'scss' },
            { name: 'less' },
            { name: 'stylus' },
            { name: 'postcss' },
            { name: 'sass' },
          ],
        },
        { name: 'scoped', valueSet: 'v' },
        { name: 'module', valueSet: 'v' },
      ],
    },
  ],
  globalAttributes: [
    {
      name: 'src',
    },
    {
      name: 'lang',
      // all other embedded languages
      values: [
        // template
        { name: 'html' },
        { name: 'pug' },
        // script
        { name: 'js' },
        { name: 'ts' },
        { name: 'jsx' },
        { name: 'tsx' },
        // style
        { name: 'css' },
        { name: 'scss' },
        { name: 'less' },
        { name: 'stylus' },
        { name: 'postcss' },
        { name: 'sass' },
        // custom block
        { name: 'md' },
        { name: 'json' },
        { name: 'jsonc' },
        { name: 'yaml' },
        { name: 'toml' },
        { name: 'gql' },
        { name: 'graphql' },
      ],
    },
  ],
}
