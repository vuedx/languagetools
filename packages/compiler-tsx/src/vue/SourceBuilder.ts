import { invariant } from '@vuedx/shared'
import type { DecodedSourceMap } from 'magic-string'
import type { SourceMapLike } from '../types/SourceMapLike'
import type { TransformedCode } from '../types/TransformedCode'

import { getMappings } from './sourceMapHelpers'

export class SourceBuilder {
  private isConsumed: boolean = false
  private code: string = ''
  private lines: number = 0

  private readonly sourceMap: DecodedSourceMap

  constructor(fileName: string, content: string) {
    this.sourceMap = {
      file: fileName,
      mappings: [],
      names: [],
      sources: [fileName],
      sourcesContent: [content],
    }
  }

  append(code: string): void
  append(code: string, sourceMap: SourceMapLike): void
  append(code: string, sourceMap?: SourceMapLike): void {
    if (this.isConsumed) throw new Error('SourceBuilder is consumed')
    const chunk = code + '\n'
    this.code += chunk
    const lines = code.split('\n').length
    this.lines += lines
    if (sourceMap != null) {
      const mappings = getMappings(sourceMap)
      invariant(mappings.length <= lines, 'Invalid source map.')
      const nameOffset = this.sourceMap.names.length
      if (sourceMap.names != null) this.sourceMap.names.push(...sourceMap.names)
      this.sourceMap.mappings.push(
        ...mappings.map((mapping) =>
          mapping.map((segment) => {
            if (segment.length === 5) {
              return [
                segment[0],
                0,
                segment[2],
                segment[3],
                segment[4] + nameOffset,
              ] as [number, number, number, number, number]
            } else if (segment.length === 4) {
              return [segment[0], 0, segment[2], segment[3]] as [
                number,
                number,
                number,
                number,
              ]
            }
            return segment
          }),
        ),
      )
    }

    for (let i = this.sourceMap.mappings.length; i < this.lines; i++) {
      this.sourceMap.mappings.push(sourceMap == null ? [[0]] : [])
    }
  }

  end(): TransformedCode {
    this.isConsumed = true
    return {
      code: this.code,
      map: this.sourceMap,
    }
  }
}
