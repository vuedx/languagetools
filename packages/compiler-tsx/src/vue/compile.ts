import {
  CompilerError,
  parse,
  SFCBlock,
  SFCDescriptor,
} from '@vuedx/compiler-sfc'
import { Cache, createCache, getComponentName } from '@vuedx/shared'
import type {
  TransformOptions,
  TransformOptionsResolved,
} from '../types/TransformOptions'
import { transformCustomBlock } from './blocks/transformCustomBlock'
import { SourceBuilder } from './SourceBuilder'
import { rebaseSourceMap } from './sourceMapHelpers'

import type { RootNode } from '@vue/compiler-core'
import type { RawSourceMap } from 'source-map'
import { encode } from 'sourcemap-codec'
import type { TransformedCode } from '../types/TransformedCode'
import { transformScript } from './blocks/transformScript'
import { transformScriptSetup } from './blocks/transformScriptSetup'
import { transformStyle } from './blocks/transformStyle'
import { transformTemplate } from './blocks/transformTemplate'

export interface CompileOptions extends TransformOptions {}

export interface CompileOutput extends TransformedCode {
  template?: RootNode
  descriptor: SFCDescriptor
  errors: Array<CompilerError | SyntaxError>
  unusedIdentifiers: string[]
}

export function compile(
  source: string,
  options: CompileOptions,
): Omit<CompileOutput, 'map'> & { map: RawSourceMap } {
  const result = compileWithDecodedSourceMap(source, options)

  return {
    ...result,
    map: {
      version: 3 as unknown as string,
      ...result.map,
      mappings: encode(result.map.mappings),
    },
  }
}

export function compileWithDecodedSourceMap(
  source: string,
  options: CompileOptions,
): CompileOutput {
  // performance.mark('beforeTransform')
  const cache = options.cache ?? createCache(100)
  const key = (name: string): string => `${options.fileName}::${name}`
  const previous = cache.get(key('descriptor')) as SFCDescriptor | undefined
  const { descriptor, errors } = parse(source)

  const lang = descriptor.script?.lang ?? descriptor.scriptSetup?.lang
  const internalIdentifierPrefix =
    options.internalIdentifierPrefix ?? '__VueDX__'
  const contextIdentifier = `${internalIdentifierPrefix}ctx`
  const typeIdentifier = `${internalIdentifierPrefix}TypeCheck`
  const resolvedOptions: TransformOptionsResolved = {
    ...options,
    runtimeModuleName: 'vue',
    typeCheckModuleName: 'vuedx~runtime',
    typeIdentifier,
    contextIdentifier,
    internalIdentifierPrefix,
    isTypeScript: options.isTypeScript ?? (lang === 'ts' || lang === 'tsx'),
    cache,
    descriptor,
  }
  const builder = new SourceBuilder(options.fileName, source)
  const template = runIfNeeded(
    key('template'),
    previous?.template,
    descriptor.template,
    cache,
    () => transformTemplate(descriptor.template, resolvedOptions),
  )

  const script = runIfNeeded(
    key('script'),
    previous?.script,
    descriptor.script,
    cache,
    () => transformScript(descriptor.script, resolvedOptions),
  )
  const scriptSetup = runIfNeeded(
    key('scriptSetup'),
    previous?.scriptSetup,
    descriptor.scriptSetup,
    cache,
    () => transformScriptSetup(descriptor.scriptSetup, resolvedOptions),
  )

  function region(name: string, fn: () => void): void {
    builder.append(`//#region ${name}`)
    fn()
    builder.append(`//#endregion`)
  }

  builder.append(
    `import * as ${resolvedOptions.typeIdentifier} from '${resolvedOptions.typeCheckModuleName}';`,
  )
  region('<script>', () => {
    builder.append(
      script.code,
      rebaseSourceMap(script.map, descriptor.script?.loc.start),
    )
  })

  if (scriptSetup != null) {
    region('<script setup>', () => {
      builder.append(
        scriptSetup.code,
        rebaseSourceMap(scriptSetup.map, descriptor.scriptSetup?.loc.start),
      )
    })
  }

  const customBlocksResults = descriptor.customBlocks.map((block, index) => {
    const result = runIfNeeded(
      key(`customBlock${index}`),
      previous?.customBlocks[index],
      block,
      cache,
      () => transformCustomBlock(block, resolvedOptions),
    )

    region(`<${block.type}>`, () => {
      builder.append(result.code, rebaseSourceMap(result.map, block.loc.start))
    })

    return result
  })

  const defaultExportIdentifier =
    scriptSetup != null ? scriptSetup.exportIdentifier : script.exportIdentifier

  builder.append(
    `const ${contextIdentifier} = ${[...customBlocksResults, script].reduce(
      (code, result) => {
        if (result.decoratorIdentifier != null) {
          return `${result.decoratorIdentifier}(${code})`
        }

        return code
      },
      `new ${defaultExportIdentifier}()`,
    )}`,
  )

  region(`<template>`, () => {
    builder.append(
      template.code,
      rebaseSourceMap(template.map, descriptor.template?.loc.start),
    )
  })

  descriptor.styles.forEach((style, index) => {
    const result = runIfNeeded(
      key(`style${index}`),
      previous?.styles[index],
      style,
      cache,
      () => transformStyle(style, resolvedOptions),
    )
    builder.append('/* <style> */')
    region('<style>', () => {
      builder.append(result.code, rebaseSourceMap(result.map, style.loc.start))
    })
  })

  region('public component definition', () => {
    builder.append(
      [
        `export default class ${getComponentName(options.fileName)} {`,
        defineProperty(
          '$props',
          `${resolvedOptions.typeIdentifier}.internal.MergeAttrs<${
            scriptSetup?.emitIdentifier != null
              ? `(typeof ${resolvedOptions.contextIdentifier}.$props & ${resolvedOptions.typeIdentifier}.internal.EmitsToProps<typeof ${scriptSetup.emitIdentifier}>)`
              : `typeof ${resolvedOptions.contextIdentifier}.$props`
          }, typeof ${template.attrsIdentifier}>`,
        ),
        defineProperty(
          '$slots',
          `${resolvedOptions.typeIdentifier}.internal.Slots<ReturnType<typeof ${template.slotsIdentifier}>>`,
        ),
        `}`,
      ].join('\n'),
    )
  })

  const output = builder.end()
  // performance.mark('afterTransform')
  // performance.measure('transform', 'beforeTransform', 'afterTransform')

  const usedIdentifiers = new Set(
    template.ast != null ? template.ast.scope.globals : [],
  )
  const identifiers =
    scriptSetup?.identifiers.filter((id) => {
      if (usedIdentifiers.has(id)) return false
      return true
    }) ?? []

  return {
    code: output.code,
    map: {
      ...output.map,
      file: options.fileName.replace(
        /\.vue$/,
        resolvedOptions.isTypeScript ? '.tsx' : '.jsx',
      ),
    },
    descriptor,
    errors: [...errors, ...template.errors],
    template: template.ast,
    unusedIdentifiers: identifiers,
  }

  function defineProperty(name: string, type: string): string {
    return resolvedOptions.isTypeScript
      ? `  ${name} = null as unknown as ${type};`
      : `  ${name} = /** @type {${type}} */ (/** @type {unknown} */ (null));`
  }
}

function runIfNeeded<Block extends SFCBlock, R>(
  key: string,
  previous: Block | undefined | null,
  current: Block | null,
  cache: Cache<string, unknown>,
  fn: () => R,
): R {
  if (cache.has(key) && hasBlockChanged(previous, current)) cache.delete(key)
  return cache.resolve(key, fn) as R
}

function hasBlockChanged(
  previous: SFCBlock | undefined | null,
  current: SFCBlock | null,
): boolean {
  if (previous === undefined) return true
  if (previous === current) return true
  if (previous == null || current == null) return false
  if (previous.type !== current.type) return false
  if (previous.lang !== current.lang) return false
  if (previous.content !== current.content) return false
  if (!areObjectsEqual(previous.attrs, current.attrs)) return false
  return true
}

function areObjectsEqual(
  a: Record<string, string | boolean>,
  b: Record<string, string | boolean>,
): boolean {
  if (Object.keys(a).length !== Object.keys(b).length) return false
  for (const key of Object.keys(a)) {
    if (a[key] !== b[key]) return false
  }

  return true
}
