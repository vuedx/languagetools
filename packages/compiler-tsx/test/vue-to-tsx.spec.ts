import FS from 'fs'
import { encode } from 'sourcemap-codec'
import Path from 'path'
import { compile, compileWithDecodedSourceMap } from '../src/vue/compile'

describe('Vue to TSX compiler', () => {
  const dir = Path.join(__dirname, 'fixtures')
  const fixtures = FS.readdirSync(dir).filter((file) => file.endsWith('.vue'))

  test.each(fixtures)('%s', async (file) => {
    const fileName = Path.join(dir, file)
    const fixture = await FS.promises.readFile(fileName, 'utf-8')
    const result = compileWithDecodedSourceMap(fixture, {
      fileName,

      isTypeScript: true,
    })
    expect(result.code).toMatchSnapshot(file)
    await writeIfChanged(fileName.replace(/\.vue$/, '.tsx'), result.code)
    await writeIfChanged(
      fileName.replace(/\.vue$/, '.tsx.map'),
      JSON.stringify({
        version: 3,
        ...result.map,
        mappings: encode(result.map.mappings),
      }),
    )
    await writeIfChanged(
      fileName.replace(/\.vue$/, '.mappings.json'),
      `[${result.map.mappings
        .map((mapping) => JSON.stringify(mapping))
        .join(',\n')}]`,
    )
  })
})

async function writeIfChanged(
  fileName: string,
  content: string,
): Promise<void> {
  const oldContent = await FS.promises
    .readFile(fileName, 'utf-8')
    .catch(() => '')
  if (oldContent !== content) {
    await FS.promises.writeFile(fileName, content)
  }
}
