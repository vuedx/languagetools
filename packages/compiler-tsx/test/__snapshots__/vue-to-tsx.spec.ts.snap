// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vue to TSX compiler if-else-if.vue: if-else-if.vue 1`] = `
"import * as __VueDX__TypeCheck from 'vuedx~runtime';
declare const __VueDX__defineComponent: typeof import("vue").defineComponent;
type __VueDX__GlobalComponents = import("vue").GlobalComponents;
//#region <script>

const __VueDX___Script_Component = __VueDX__defineComponent({});
//#endregion
//#region <script setup>

;const __VueDX___ScriptSetup_scope = __VueDX__TypeCheck.internal.scope(async () => {
const __VueDX___ScriptSetup_props = defineProps({});
const __VueDX___ScriptSetup_emits = ({});
const __VueDX___ScriptSetup_expose = {};
const __VueDX___ScriptSetup_internalProps = {};
const __VueDX___ScriptSetup_Component = __VueDX__defineComponent((_: typeof __VueDX___ScriptSetup_internalProps)=> {});

//#endregion
function __VueDX__RegisterSelf<T>(ctx: T) {
  return { ...ctx, ["IfElseIf"]: IfElseIf }
}
const __VueDX__ctx = __VueDX__RegisterSelf(new __VueDX___Script_Component())
//#region <template>
function __VueDX__render() {
  /*<vuedx:templateGlobals>*/
  let type = __VueDX__ctx.type
  let $slots = __VueDX__ctx.$slots
  /*</vuedx:templateGlobals>*/
  return (
    <>
      {type === 'A'
        ? <>
            <div
              /*<vuedx:tsx-completions-target/>*/
            >
              {"A"}
            </div>
          </>
        : type === 'B'
        ? <>
            <div
              /*<vuedx:tsx-completions-target/>*/
            >
              {"B"}
            </div>
          </>
        : type === 'C'
        ? <>
            <div
              /*<vuedx:tsx-completions-target/>*/
            >
              {"C"}
            </div>
          </>
        : <>
            <div
              /*<vuedx:tsx-completions-target/>*/
            >
              {"Not A/B/C"}
            </div>
          </>
      }
    </>
  )
}
__VueDX__render();
/*<vuedx:diagnosticsIgnore>*/
function __VueDX___slots() {
  /*<vuedx:templateGlobals>*/
  let type = __VueDX__ctx.type
  let $slots = __VueDX__ctx.$slots
  /*</vuedx:templateGlobals>*/
  return __VueDX__TypeCheck.internal.union(__VueDX__TypeCheck.internal.flat([
  ]))
}
/*</vuedx:diagnosticsIgnore>*/
const __VueDX___attrs = (() => {
  return __VueDX__TypeCheck.internal.first(__VueDX__TypeCheck.internal.flat([
    {} as unknown as __VueDX__TypeCheck.internal.PropsOf<JSX.IntrinsicElements, "div">,
    {} as unknown as __VueDX__TypeCheck.internal.PropsOf<JSX.IntrinsicElements, "div">,
    {} as unknown as __VueDX__TypeCheck.internal.PropsOf<JSX.IntrinsicElements, "div">,
    {} as unknown as __VueDX__TypeCheck.internal.PropsOf<JSX.IntrinsicElements, "div">,
  ]))
})();
//#endregion
return {__VueDX___ScriptSetup_Component, __VueDX___ScriptSetup_props, __VueDX___ScriptSetup_emits, __VueDX___ScriptSetup_expose, __VueDX___attrs, __VueDX___slots, __VueDX__ctx};});
const {__VueDX___ScriptSetup_Component, __VueDX___ScriptSetup_props, __VueDX___ScriptSetup_emits, __VueDX___ScriptSetup_expose, __VueDX___attrs, __VueDX___slots, __VueDX__ctx} = __VueDX___ScriptSetup_scope;
//#region public component definition
const IfElseIfPublic = null as unknown as new () => typeof __VueDX___ScriptSetup_expose;
export default class IfElseIf extends IfElseIfPublic {
  $props = null as unknown as __VueDX__TypeCheck.internal.MergeAttrs<typeof __VueDX__ctx.$props, typeof __VueDX___attrs>;
  $slots = null as unknown as __VueDX__TypeCheck.internal.Slots<ReturnType<typeof __VueDX___slots>>;
}
//#endregion
"
`;

exports[`Vue to TSX compiler nested-for-slot.vue: nested-for-slot.vue 1`] = `
"import * as __VueDX__TypeCheck from 'vuedx~runtime';
declare const __VueDX__defineComponent: typeof import("vue").defineComponent;
type __VueDX__GlobalComponents = import("vue").GlobalComponents;
//#region <script>

const __VueDX___Script_Component = __VueDX__defineComponent({});
//#endregion
//#region <script setup>

;const __VueDX___ScriptSetup_scope = __VueDX__TypeCheck.internal.scope(async () => {
const __VueDX___ScriptSetup_props = defineProps({});
const __VueDX___ScriptSetup_emits = ({});
const __VueDX___ScriptSetup_expose = {};
const __VueDX___ScriptSetup_internalProps = {};
const __VueDX___ScriptSetup_Component = __VueDX__defineComponent((_: typeof __VueDX___ScriptSetup_internalProps)=> {});

//#endregion
function __VueDX__RegisterSelf<T>(ctx: T) {
  return { ...ctx, ["NestedForSlot"]: NestedForSlot }
}
const __VueDX__ctx = __VueDX__RegisterSelf(new __VueDX___Script_Component())
//#region <template>
function __VueDX__render() {
  /*<vuedx:templateGlobals>*/
  let list = __VueDX__ctx.list
  let props = __VueDX__ctx.props
  let $slots = __VueDX__ctx.$slots
  /*</vuedx:templateGlobals>*/
  return (
    <>
      {
        __VueDX__TypeCheck.internal.renderList(list, (items) => {
          return (
            <div
              /*<vuedx:tsx-completions-target/>*/
            >
              {
                $slots.default != null ? $slots.default({}) : (
                  <>
                    {"fallback content"}
                  </>
                )

              }
              {
                __VueDX__TypeCheck.internal.renderList(items, (item) => {
                  return (
                    <span
                      /*<vuedx:tsx-completions-target/>*/
                    >
                      {
                        $slots.other != null ? $slots.other({
                          ...(props),
                          myProp: item,
                        }) : (
                          <>
                            {"fallback content"}
                          </>
                        )

                      }
                    </span>
                  )
                })
              }
            </div>
          )
        })
      }
    </>
  )
}
__VueDX__render();
/*<vuedx:diagnosticsIgnore>*/
function __VueDX___slots() {
  /*<vuedx:templateGlobals>*/
  let list = __VueDX__ctx.list
  let props = __VueDX__ctx.props
  let $slots = __VueDX__ctx.$slots
  /*</vuedx:templateGlobals>*/
  return __VueDX__TypeCheck.internal.union(__VueDX__TypeCheck.internal.flat([
    __VueDX__TypeCheck.internal.flat(
      __VueDX__TypeCheck.internal.renderList(list, (items) => (
        {
          default: {},
        }
      ))
    ),
    __VueDX__TypeCheck.internal.flat(
      __VueDX__TypeCheck.internal.renderList(list, (items) => (
        __VueDX__TypeCheck.internal.flat(
          __VueDX__TypeCheck.internal.renderList(items, (item) => (
            {
              "other": {
                ...(props),
                myProp: item,
              },
            }
          ))
        )
      ))
    ),
  ]))
}
/*</vuedx:diagnosticsIgnore>*/
const __VueDX___attrs = (() => {
  return __VueDX__TypeCheck.internal.first(__VueDX__TypeCheck.internal.flat([
    {} as unknown as __VueDX__TypeCheck.internal.PropsOf<JSX.IntrinsicElements, "div">,
  ]))
})();
//#endregion
return {__VueDX___ScriptSetup_Component, __VueDX___ScriptSetup_props, __VueDX___ScriptSetup_emits, __VueDX___ScriptSetup_expose, __VueDX___attrs, __VueDX___slots, __VueDX__ctx};});
const {__VueDX___ScriptSetup_Component, __VueDX___ScriptSetup_props, __VueDX___ScriptSetup_emits, __VueDX___ScriptSetup_expose, __VueDX___attrs, __VueDX___slots, __VueDX__ctx} = __VueDX___ScriptSetup_scope;
//#region public component definition
const NestedForSlotPublic = null as unknown as new () => typeof __VueDX___ScriptSetup_expose;
export default class NestedForSlot extends NestedForSlotPublic {
  $props = null as unknown as __VueDX__TypeCheck.internal.MergeAttrs<typeof __VueDX__ctx.$props, typeof __VueDX___attrs>;
  $slots = null as unknown as __VueDX__TypeCheck.internal.Slots<ReturnType<typeof __VueDX___slots>>;
}
//#endregion
"
`;

exports[`Vue to TSX compiler ts-script-template.vue: ts-script-template.vue 1`] = `
"import * as __VueDX__TypeCheck from 'vuedx~runtime';
declare const __VueDX__defineComponent: typeof import("vue").defineComponent;
type __VueDX__GlobalComponents = import("vue").GlobalComponents;
//#region <script>

import { defineComponent } from 'vue'
const __VueDX___Script_Component =  defineComponent({
  data: () => ({ foo: 'string' }),
})
//#endregion
//#region <script setup>

;const __VueDX___ScriptSetup_scope = __VueDX__TypeCheck.internal.scope(async () => {
const __VueDX___ScriptSetup_props = defineProps({});
const __VueDX___ScriptSetup_emits = ({});
const __VueDX___ScriptSetup_expose = {};
const __VueDX___ScriptSetup_internalProps = {};
const __VueDX___ScriptSetup_Component = __VueDX__defineComponent((_: typeof __VueDX___ScriptSetup_internalProps)=> {});

//#endregion
function __VueDX__RegisterSelf<T>(ctx: T) {
  return { ...ctx, ["TsScriptTemplate"]: TsScriptTemplate }
}
const __VueDX__ctx = __VueDX__RegisterSelf(new __VueDX___Script_Component())
//#region <template>
function __VueDX__render() {
  /*<vuedx:templateGlobals>*/
  let foo = __VueDX__ctx.foo
  let console = __VueDX__ctx.console
  let $slots = __VueDX__ctx.$slots
  /*</vuedx:templateGlobals>*/
  return (
    <>
      <div

        class={{ foo: foo }}
        innerHTML={foo}
        {...({[foo]: true})}
        {...(foo)}
        onClick={__VueDX__TypeCheck.internal.first([
          () => {
          console.log
          }, 
          () => {
          console.log
          }, 
          () => {
          console.log
          }, 
        ])}
        /*<vuedx:tsx-completions-target/>*/
      >
        {foo}
      </div>
    </>
  )
}
__VueDX__render();
/*<vuedx:diagnosticsIgnore>*/
function __VueDX___slots() {
  /*<vuedx:templateGlobals>*/
  let foo = __VueDX__ctx.foo
  let console = __VueDX__ctx.console
  let $slots = __VueDX__ctx.$slots
  /*</vuedx:templateGlobals>*/
  return __VueDX__TypeCheck.internal.union(__VueDX__TypeCheck.internal.flat([
  ]))
}
/*</vuedx:diagnosticsIgnore>*/
const __VueDX___attrs = (() => {
  return __VueDX__TypeCheck.internal.first(__VueDX__TypeCheck.internal.flat([
    {} as unknown as __VueDX__TypeCheck.internal.PropsOf<JSX.IntrinsicElements, "div">,
  ]))
})();
//#endregion
return {__VueDX___ScriptSetup_Component, __VueDX___ScriptSetup_props, __VueDX___ScriptSetup_emits, __VueDX___ScriptSetup_expose, __VueDX___attrs, __VueDX___slots, __VueDX__ctx};});
const {__VueDX___ScriptSetup_Component, __VueDX___ScriptSetup_props, __VueDX___ScriptSetup_emits, __VueDX___ScriptSetup_expose, __VueDX___attrs, __VueDX___slots, __VueDX__ctx} = __VueDX___ScriptSetup_scope;
//#region public component definition
const TsScriptTemplatePublic = null as unknown as new () => typeof __VueDX___ScriptSetup_expose;
export default class TsScriptTemplate extends TsScriptTemplatePublic {
  $props = null as unknown as __VueDX__TypeCheck.internal.MergeAttrs<typeof __VueDX__ctx.$props, typeof __VueDX___attrs>;
  $slots = null as unknown as __VueDX__TypeCheck.internal.Slots<ReturnType<typeof __VueDX___slots>>;
}
//#endregion
"
`;
